ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java

```diff
@@ -52,27 +50,18 @@ private EJBValidationConfigurator() {
 
     @Override
     public void configure(final DeploymentPhaseContext context, final ComponentDescription description, final ComponentConfiguration configuration) throws DeploymentUnitProcessingException {
-        final DeploymentReflectionIndex deploymentReflectionIndex = context.getDeploymentUnit().getAttachment(Attachments.REFLECTION_INDEX);
-        final ClassReflectionIndex classIndex = deploymentReflectionIndex.getClassIndex(configuration.getComponentClass());
+        final ClassReflectionIndex classIndex = context.getDeploymentUnit().getAttachment(Attachments.REFLECTION_INDEX).getClassIndex(configuration.getComponentClass());
         final Constructor<?> ctor = classIndex.getConstructor(new String[0]);
         boolean noInterface = false;
         for(ViewDescription view : description.getViews()) {
             if(view.getViewClassName().equals(description.getComponentClassName())) {
                 noInterface = true;
             }
         }
-        if(ctor == null && noInterface) {
-            //we only validate this for no interface views
-            throw EjbLogger.ROOT_LOGGER.ejbMustHavePublicDefaultConstructor(description.getComponentName(), description.getComponentClassName());
-        }
-        if(configuration.getComponentClass().getEnclosingClass() != null) {
-            throw EjbLogger.ROOT_LOGGER.ejbMustNotBeInnerClass(description.getComponentName(), description.getComponentClassName());
-        }
-        if(!Modifier.isPublic(configuration.getComponentClass().getModifiers())) {
-            throw EjbLogger.ROOT_LOGGER.ejbMustBePublicClass(description.getComponentName(), description.getComponentClassName());
-        }
-        if(Modifier.isFinal(configuration.getComponentClass().getModifiers())) {
-            throw EjbLogger.ROOT_LOGGER.ejbMustNotBeFinalClass(description.getComponentName(), description.getComponentClassName());
-        }
+
+        EjbValidationsUtil.verifyEjbClassAndDefaultConstructor(ctor, configuration.getComponentClass().getEnclosingClass(), noInterface, description.getComponentName(), description.getComponentClassName(), configuration.getComponentClass().getModifiers());
+        EjbValidationsUtil.verifyEjbMethodAreNotFinalNorStatic(configuration.getComponentClass().getDeclaredMethods(),description.getComponentClassName());
```

Comment:
@rpelisse @mjurc The Jira states: "The ejb 3.2 core specification state in chapter 4.9.6 that business methods should not be declared as final or static.". Since we're doing class.getDeclaredMethods() here we are getting all methods of the class itself, and verifying those. To me that looks wrong as getDeclaredMethods() will return all private methods as well, which are not business methods. Also, does a method being public necessarily need to be a business method (e.g. if a local/remote interfaces are defined and the method is not declared there)? Finally, the ejb class could inherit some business methods from a super class, so we should probably check those as well?

Reply:
>  To me that looks wrong as getDeclaredMethods() will return all private methods as well, which are not business methods. Also, does a method being public necessarily need to be a business method (e.g. if a local/remote interfaces are defined and the method is not declared there)?

I've added some logic to adress that.

> Finally, the ejb class could inherit some business methods from a super class, so we should probably check those as well?

I would not got that far. I think validation should take the low hanging fruit, but starting to explore inheritance and such things sounds a bit like overkill. Also, I would rather introduce the current set of changes, see if it breaks and leads to issues and then maybe go deeper and try to validate more advanced use case. (IMHO)